File: frontend/src/lib/services/knowledge.ts
import { api } from "../api-client";
import type { Paginated } from "./common";

/**
 * 知识库服务：严格对齐 docs/API.md §6。
 * 字段定义权在 docs/DATABASE.md knowledge_documents 表;前端不得假设契约之外的字段。
 */

/** 文档索引状态枚举(定义权:docs/DATABASE.md knowledge_documents.status)。 */
export type KnowledgeDocumentStatus =
  | "uploaded"
  | "parsing"
  | "indexing"
  | "indexed"
  | "failed"
  | "outdated";

export const KNOWLEDGE_DOCUMENT_STATUSES: KnowledgeDocumentStatus[] = [
  "uploaded",
  "parsing",
  "indexing",
  "indexed",
  "failed",
  "outdated",
];

/** 文档列表元素(API.md §6)。 */
export type KnowledgeDocument = {
  id: string;
  title: string;
  document_type: string | null;
  status: KnowledgeDocumentStatus;
  version: string | null;
  issuing_authority: string | null;
  effective_date: string | null;
  expiration_date: string | null;
  chunk_count: number | null;
  created_at: string;
  updated_at: string;
};

/** 列表查询参数(page 默认 1,page_size 默认 20;status 可选过滤)。 */
export type KnowledgeListParams = {
  page?: number;
  page_size?: number;
  status?: KnowledgeDocumentStatus;
};

/** POST /api/knowledge/documents 响应:解析与索引为异步任务(knowledge_indexing)。 */
export type KnowledgeUploadResponse = {
  document_id: string;
  task_id: string;
};

/** DELETE /api/knowledge/documents/{id} 响应。 */
export type KnowledgeDeleteResponse = {
  id: string;
  deleted: boolean;
};

/** POST /api/knowledge/rebuild 响应:全量重建为异步任务(knowledge_reindexing)。 */
export type KnowledgeRebuildResponse = {
  task_id: string;
};

/** GET /api/knowledge/status 响应:知识库聚合计数;尚无文档时计数为 0、last_indexed_at 为 null。 */
export type KnowledgeStatus = {
  document_count: number;
  indexed_count: number;
  indexing_count: number;
  failed_count: number;
  last_indexed_at: string | null;
};

const base = "/api/knowledge";

export const knowledgeService = {
  list: (params: KnowledgeListParams = {}, signal?: AbortSignal) =>
    api.get<Paginated<KnowledgeDocument>>(`${base}/documents`, {
      query: { page: params.page, page_size: params.page_size, status: params.status },
      signal,
    }),
  /** 上传知识库源文档(multipart/form-data,字段 file);类型与大小限制见 API.md §9。 */
  upload: (form: FormData) => api.post<KnowledgeUploadResponse>(`${base}/documents`, form),
  /** 软删除文档并移除其索引数据。 */
  remove: (id: string) =>
    api.delete<KnowledgeDeleteResponse>(`${base}/documents/${encodeURIComponent(id)}`),
  /** 触发全量索引重建(异步任务)。 */
  rebuild: () => api.post<KnowledgeRebuildResponse>(`${base}/rebuild`),
  /** 知识库聚合计数。 */
  status: (signal?: AbortSignal) => api.get<KnowledgeStatus>(`${base}/status`, { signal }),
};

File: backend/app/api/routers/knowledge.py
"""知识库路由（API.md §6）。保持薄：解析请求、调用 KnowledgeBaseService。

权限（specs/knowledge-base.md）：上传/删除/重建仅 admin；查询类端点所有认证用户。
"""

import uuid

from fastapi import APIRouter, Depends, File, Header, Query, Request, UploadFile

from app.api.dependencies import (
    CurrentUser,
    DbSession,
    get_request_id,
    require_permission,
)
from app.core.cache import PREFIX_KNOWLEDGE_STATUS, get_cache
from app.core.config import get_settings
from app.models.user import User
from app.schemas.common import Page
from app.schemas.knowledge import (
    DocumentStatus,
    KnowledgeDeleteResponse,
    KnowledgeDocumentListItem,
    KnowledgeRebuildResponse,
    KnowledgeStatusResponse,
    KnowledgeUploadResponse,
)
from app.services.knowledge_service import KnowledgeBaseService

router = APIRouter(prefix="/knowledge", tags=["knowledge"])

AdminUser = Depends(require_permission("knowledge.manage"))


def _to_list_item(doc) -> KnowledgeDocumentListItem:
    return KnowledgeDocumentListItem(
        id=doc.id,
        title=doc.title,
        document_type=doc.document_type,
        status=doc.status,
        version=doc.version,
        issuing_authority=doc.issuing_authority,
        effective_date=doc.effective_date,
        chunk_count=doc.chunk_count,
        created_at=doc.created_at,
        updated_at=doc.updated_at,
    )


@router.get("/documents", response_model=Page[KnowledgeDocumentListItem])
def list_documents(
    session: DbSession,
    current_user: CurrentUser,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    status: DocumentStatus | None = None,
) -> Page[KnowledgeDocumentListItem]:
    rows, total = KnowledgeBaseService(session).list_documents(
        str(status) if status else None, page, page_size
    )
    return Page(
        items=[_to_list_item(r) for r in rows],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post("/documents", response_model=KnowledgeUploadResponse)
def upload_document(
    session: DbSession,
    request: Request,
    current_user: User = AdminUser,
    file: UploadFile = File(...),
    idempotency_key: str | None = Header(None),
) -> KnowledgeUploadResponse:
    document, task = KnowledgeBaseService(session).upload(
        user=current_user,
        filename=file.filename,
        content_type=file.content_type,
        data=file.file.read(),
        request_id=get_request_id(request),
        idempotency_key=idempotency_key,
    )
    return KnowledgeUploadResponse(document_id=document.id, task_id=task.id)


@router.delete("/documents/{document_id}", response_model=KnowledgeDeleteResponse)
def delete_document(
    document_id: uuid.UUID,
    session: DbSession,
    request: Request,
    current_user: User = AdminUser,
) -> KnowledgeDeleteResponse:
    document = KnowledgeBaseService(session).delete(
        current_user, document_id, request_id=get_request_id(request)
    )
    return KnowledgeDeleteResponse(id=document.id, deleted=True)


@router.post("/rebuild", response_model=KnowledgeRebuildResponse)
def rebuild_index(
    session: DbSession,
    request: Request,
    current_user: User = AdminUser,
    idempotency_key: str | None = Header(None),
) -> KnowledgeRebuildResponse:
    task = KnowledgeBaseService(session).rebuild(
        current_user,
        request_id=get_request_id(request),
        idempotency_key=idempotency_key,
    )
    return KnowledgeRebuildResponse(task_id=task.id)


@router.get("/status", response_model=KnowledgeStatusResponse)
def knowledge_status(session: DbSession, current_user: CurrentUser) -> dict:
    # 全库共享的只读聚合：进程内 TTL 缓存（M7），上传/删除/重建/索引终态后失效
    cached = get_cache().get(PREFIX_KNOWLEDGE_STATUS)
    if isinstance(cached, dict):
        return cached
    result = KnowledgeBaseService(session).status()
    get_cache().set(PREFIX_KNOWLEDGE_STATUS, result, get_settings().CACHE_TTL_SECONDS)
    return result
